### ISSUE TITLE
PostgreSQL support

### ISSUE THREAD

**Issue by navossoc** [hardware/mailserver#222]
#### Classification

* [X] Question
* [x] Feature/Enhancement

#### Description
Hi, I would like to know if you have any plans to replace MariaDB by PostgreSQL or maybe add it as an second database option.

Sure, maybe you are not willing to do it, but, may you accept some pull request with it?
Let me know.

[]'s

**Comment by hardware** [comment]
> Sure, maybe you are not willing to do it, but, may you accept some pull request with it?

Sure, go for it, contributions are welcome.

**Comment by navossoc** [comment]
To be honest I already did it (replaced), I'm testing it for a quite while.

Do you prefer to change the database or add as a second option?
If so, how you want to make this option available? Any thoughts?

**Comment by hardware** [comment]
> Do you prefer to change the database or add as a second option ?

With a second option in a backwards-compatible manner.

> If so, how you want to make this option available ?

You can add a new environment variable called `DBDRIVER` with two options : `mariadb` and `postgresql`. Use gucci and [Golang templating functions](https://golang.org/pkg/text/template/) to setup all configuration files needed. 

Example :

```bash
# /etc/dovecot/dovecot-sql.conf.ext

{{ if eq .DBDRIVER "mariadb" }}
    driver = mysql
{{ else if eq .DBDRIVER "postgresql" }} 
    driver = pgsql
{{ end }}

connect = host={{ .DBHOST }} port={{ .DBPORT }} dbname={{ .DBNAME }} user={{ .DBUSER }} password={{ .DBPASS }}
[...]
```

### Testing 

When all your changes are ready for testing, complete the makefile :

* Add [PostgreSQL](https://hub.docker.com/_/postgres/) official docker image 
* Use `mailserver_reverse` to test PostgreSQL support

```bash
# Makefile
# https://github.com/hardware/mailserver/blob/master/Makefile

docker run \
        -d \
	--name postgresql \
	-e POSTGRES_PASSWORD=testpasswd \
	-e POSTGRES_USER=postfix \
	-v "`pwd`/test/config/postgresql":/docker-entrypoint-initdb.d \
	-t postgres:10

docker run \
        -d \
        --name mailserver_reverse \
        --link postgresql:postgresql \
        -e DBDRIVER=postgresql \
        -e DBHOST=postgresql \
        -e DBPASS=/tmp/passwd/postgresql \
        ...
```

* Export a basic SQL structure in `/test/config/postgresql` folder. This file is called before starting the service to initialize the database. More info : https://hub.docker.com/_/postgres/

Your SQL structure must contain 1 domain, 2 mailboxes, 4 alias, 2 quotas :

- domain : **domain.tld**
- username : **sarah.connor@domain.tld**, password : **testpasswd12**
- username : **john.doe@domain.tld**, password : **testpasswd12**

```sql
INSERT INTO `domain` VALUES ('ALL','',0,0,0,0,'',0,'2016-11-28 08:53:31','2016-11-28 08:53:31',1),('domain.tld','Test domain',0,0,1,0,'virtual',0,'2016-11-28 08:54:26','2016-11-28 08:54:26',1);

INSERT INTO `mailbox` VALUES ('john.doe@domain.tld','{SHA512-CRYPT}$6$v1LkarodHyGGmfoy$ZszVBzfEZ0CaVnYaBasgvaHJUCNfxwD/E0eNy3iuix56Vl1ZcuDvG9PVr9JRZx5k.7wp1nMb5M1V4aZXo2yfn0','John DOE','domain.tld/john.doe/',1024000,'john.doe','domain.tld','2016-11-28 08:56:47','2016-11-28 08:56:47',1),('sarah.connor@domain.tld','{SHA512-CRYPT}$6$ub.zCcyeaM7Mhs6S$rL4Yj2.Zsk8aFoF5l1mAddVrPo.UZ/1UrNwBC7UTBrX47cViSHo5eepEes6jMqC21P3cBm82adqJZvo91Ekme0','Sarah CONNOR','domain.tld/sarah.connor/',1024000,'sarah.connor','domain.tld','2016-11-28 08:57:51','2016-11-28 08:57:51',1);

INSERT INTO `alias` VALUES ('postmaster@domain.tld','john.doe@domain.tld','domain.tld','2016-11-28 08:54:26','2016-11-28 08:58:19',1),('hostmaster@domain.tld','john.doe@domain.tld','domain.tld','2016-11-28 08:54:26','2016-11-28 08:58:19',1),('john.doe@domain.tld','john.doe@domain.tld','domain.tld','2016-11-28 08:56:47','2016-11-28 08:56:47',1),('sarah.connor@domain.tld','sarah.connor@domain.tld','domain.tld','2016-11-28 08:57:51','2016-11-28 08:57:51',1);

INSERT INTO `quota2` VALUES ('john.doe@domain.tld',0,0),('sarah.connor@domain.tld',0,0);
```

When everything is ready, run `make all` and check if all tests pass.

**Comment by navossoc** [comment]
Okay, all this I had already done, including the tests.
I'm working now to support both databases (including the conditionals).

I'm doing the changes smaller, to make it very smooth to review.

Just two things I want know:

1. Currently we have a container called `mariadb`.
Should we add another container called `postgres` or go generic and call it `sqlserver`?

For compatible reasons, I think we should add a new one, but that brings a new problem.

"How" can you choose between `maria` or `postgres` container on `docker-compose.sample.yml`?
Should I leave the container config under mariadb and let the user comment/uncomment it and make the other appropriated changes?

**For example:**
```yml
  # Database
  # https://github.com/docker-library/mariadb
  # https://mariadb.org/
  mariadb:
    image: mariadb:10.2
    container_name: mariadb
    restart: ${RESTART_MODE}
    # Info : These variables are ignored when the volume already exists (if databases was created before).
    environment:
      - MYSQL_RANDOM_ROOT_PASSWORD=yes
      - MYSQL_DATABASE=postfix
      - MYSQL_USER=postfix
      - MYSQL_PASSWORD=${DATABASE_USER_PASSWORD}
    volumes:
      - ${VOLUMES_ROOT_PATH}/mysql/db:/var/lib/mysql
    networks:
      - mail_network

  # Database
  # https://github.com/docker-library/postgres
  # https://postgresql.org/
  #postgres:
  #  image: postgres:10.3
  #  container_name: postgres
  #  restart: ${RESTART_MODE}
  #  # Info : These variables are ignored when the volume already exists (if databases was created before).
  #  environment:
  #    - POSTGRES_DB=postfix
  #    - POSTGRES_USER=postfix
  #    - POSTGRES_PASSWORD=${DATABASE_USER_PASSWORD}
  #  volumes:
  #    - ${VOLUMES_ROOT_PATH}/pgsql/db:/var/lib/postgresql/data
  #  networks:
  #    - mail_network
```

I think this may solve the problem, but again, we have more containers on that compose file.
So all that others `depends_on` should be in sync.

```yml
    depends_on:
      - mailserver
      - mariadb
    # - postgres
```

If the container was called `sqlserver` (or whatever) that will not be a problem.

Maybe should we create two `docker-compose.sample.yml`?
Any thoughts about this?

2. And... How about the tests?
Should `make all` test both mariadb and postgresql or create individual tests?

Again, we will have the problem with the `--link` argument.

I didn't look too much at the tests.
I know they are passing, but I'm not sure how they really work under the hood.

**Comment by hardware** [comment]
1. The default dbms is MariaDB and for compatibility reason, it should stay that way. I advise you to add a new section in the readme describing how to use PostgreSQL instead of MariaDB. The `docker-compose.sample.yml` file and default values of `DBHOST` & `DBPORT` should remains unchanged.

2. `make all` must run both MariaDB and PostgreSQL tests with two separate containers (`mailserver_default` for MariaDB, `mailserver_reverse` for PostgreSQL). But don't worry, I'll add tests myself when reviewing your pull request. On your side, you just have to check that all tests pass successfully.

**Comment by denji** [comment]
@hardware How about support __postfixadmin__ & __Fetchmail__ `DBDRIVER=postgresql`

**Comment by navossoc** [comment]
@denji I already covered it here...

postfixadmin and rainloop are much more simpler to adapt.

**Comment by navossoc** [comment]
Well, I think mailserver is done.

I did a good cleanup here on my system before running the tests.

```shell
docker system prune -a -f --volumes
git clone https://github.com/navossoc/mailserver.git
cd mailserver
git submodule update --init
```

**mysql driver tests:**
```shell
make all
```

> 191 tests, 0 failures

**pgsql driver tests:**
```shell
make -f Makefile.postgres all
```

> 191 tests, 0 failures

---

You can check the changes here:
https://github.com/navossoc/mailserver

About the `DBDRIVER` environment variable:
I have opted by `mysql` and `pgsql`, because these are the names used by postfix/dovecot database adapters.
http://www.postfix.org/mysql_table.5.html
http://www.postfix.org/pgsql_table.5.html

Sure, you are using MariaDB, but postfix/dovecot adapters are built for `mysql`.
So, for now it is compatible with MariaDB, but we don't know in the future if they will break the compatibility and create a `mariadb` adapter.

I think this way will be better if you want to add another database (SQLite, LDAP?) on the future too.

I'll make `postfixadmin` and `rainloop` after my lunch.

Let me know if you want to change something.

**Comment by hardware** [comment]
Very clean, nice work :)

Be cureful, for postfixadmin, ` $CONF['database_port']` only works with PostgreSQL driver. David Goodwin added the MySQLi port support in september 2017 but this change is not yet available in the latest version (3.1).

* https://github.com/postfixadmin/postfixadmin/commit/4b999b3f6bb78f595af2a7294f6b231de8f61fc4
* https://github.com/hardware/postfixadmin/issues/9

**Comment by navossoc** [comment]
@hardware Thanks for the tip. I'll check it later.

Something came up now.
I'll probably end up at night.

[]'s

**Comment by navossoc** [comment]
I'm back...
Well I see what you mean about the version 3.1.

Do you think we should add the option now and let a warning (that isn't working yet) for MySQL/MariaDB users?

In my opinion we can do that, when they release postfix 3.2 or something like that, everything will be ready. 
You will just need to remove the warning.

I don't see any problems doing that.

**Comment by navossoc** [comment]
I think I'm done!

https://github.com/navossoc/mailserver
https://github.com/navossoc/postfixadmin
https://github.com/navossoc/rainloop

I'm sending the pull requests now...
Feel free to change whatever you want.

You have more in-depth of the project, so, you will make better decisions.

**Comment by hardware** [comment]
Thank you for your work. I will review your pull requests after work. At first glance, everything seems good to me.

**Comment by hardware** [comment]
@navossoc PostgreSQL support has been merged to mailserver master branch and added to postfixadmin and rainloop docker images. If everything works as expected, I will merge PostgreSQL support on the mailserver stable branch in 2 weeks.

Thanks again for your amazing work.
